### PR TITLE
wayland: Support wp-fractional-scale-v1 protocol

### PR DESCRIPTION
**Pull request by davidedmundson**
A pending protocol adds support for more native communication of
fractional scaling.

Everything in the wayland backend already existed only the fractional
scale was calculated implicitly through a combination of output size
guesswork.

The new protocol makes that explicit. I have left the old code in place
for now.

draft because:
 1) it needs more testing. Volunteers please!
2) the protocol is not yet landed, which is a chicken and egg with
approval here

### CODE REVIEW

**Comment by Kontrabant** [comment]
Looks good at first glance to me as well.  This protocol has been wanted for a while by a lot of people, so hopefully it will be finalized sooner rather than later.

**Comment by davidedmundson** [comment]
Link to upstream wayland discussions: https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/143

**Comment by flibitijibibo** [comment]
Marked this as 3.0, but if by some chance the protocol gets merged before tag this should be a really safe patch to roll in.

**Comment by flibitijibibo** [comment]
The protocol was approved this morning, so marking this as ready for review. @davidedmundson: Should we merge this after the protocol is merged or after the version is tagged? Do we expect any changes to be made in between those milestones?

**Comment by davidedmundson** [comment]
> The protocol was approved this morning, so marking this as ready for review. 

I forgot to post that I finally did some more extensive testing. I have one tiny issue with fullscreen to follow up on where the older path and this conflict in one specific chain. I'll sort it soon. 

@davidedmundson: Should we merge this after the protocol is merged or after the version is tagged? Do we expect any changes to be made in between those milestones?

It's extremely unlikely to change, but if SDL got released with this and then there was a binary incompatiable modification to the protocol before wayland-protocols release the ramifications would be huge. For that reason I would err on the side of waiting.

**Comment by Kontrabant** [comment]
Plus, it's going to take a little while for compositors to roll out new stable versions with support for it and the next SDL release is a way off, so there's no real reason to rush it through.  They seem to tag new releases pretty quickly when a new protocol is approved anyways.

> I forgot to post that I finally did some more extensive testing. I have one tiny issue with fullscreen to follow up on where the older path and this conflict in one specific chain. I'll sort it soon.

What's the issue, if you don't mind me asking?

File: src/video/wayland/SDL_waylandwindow.c
Comment (Kontrabant) [inline src/video/wayland/SDL_waylandwindow.c:1103]:
```suggestion
```

This needs to be removed here...

File: src/video/wayland/SDL_waylandwindow.c
Comment (Kontrabant) [inline src/video/wayland/SDL_waylandwindow.c:1106]:
```suggestion
    if (!window->fractional_scale) {
        update_scale_factor(window);
    }
```

and guarded here as fullscreen window moves have to be handled before updating the scale factor.

**Review by Kontrabant — COMMENTED** [review]
Little fix due to a change since the first version of this patch, but otherwise still LGTM.

File: src/video/wayland/SDL_waylandwindow.c
Comment (davidedmundson) [reply src/video/wayland/SDL_waylandwindow.c:1106]: > I forgot to post that I finally did some more extensive testing. I have one tiny issue with fullscreen to follow up on where the older path and this conflict in one specific chain. I'll sort it soon.

It turns out this is a pre-existing issue, I can reproduce it on master.  I'll file a separate bug so we don't derail this.

**Review by flibitijibibo — APPROVED** [review]
Latest lgtm. Once the final xml is tagged in a release we should be able to merge this into both SDL3 and SDL2!

**Comment by flibitijibibo** [comment]
Protocol got merged this morning! @davidedmundson, if you have time: Definitely make an updated commit with the final commit/protocol text, and as soon as the tag is out I should be able to merge (assuming nothing changes between now and tag).

**Comment by Kontrabant** [comment]
wayland-protocols 1.31 is now tagged and officially released, so everything is ready.

**Comment by davidedmundson** [comment]
I filed a bug on the existing issue I found: https://github.com/libsdl-org/SDL/issues/6707 but as I said it's pre-existing.

I'll try and find time to fix it if no-one beats me to it.

**Pull request merged**
